File: projects/03-rabinKarp/implementations/rabinKarp.cpp
#include <iostream> 
#include <string>

using namespace std;

// d is the number possible characters (1 char is 8 bits so 2^8=256 possibilities)
int d = 256;
// prime number to create a better hash function
int q = 101;

// simple function that checks if p and s.substring(start, start+p.length) are equal
bool check_strings_equal (string s, string p, int start) {
    for (int j = 0; j < p.length(); j++) {
        if (s[start+j] != p[j]) {
            return false;
        }
    }
    return true;
}

// Given a string s, and a pattern to find in the string p, this function will return the first index of the pattern p in the string s
// If the pattern is not found in s or the length of p is greater than s, the function returns -1
// we use the standard equation for the rolling hash h = (d * h + s[i]) % q;
int rabin_karp(string s, string p) {
    // get the length of s and p for later use, as well as ensuring s is longer than p
    int len_p = p.length();
    int len_s = s.length();
    if (len_p > len_s) return -1;

    // represents the constant d^(len_p-1), used to remove the left-most character from hash_s
    int d_p_1 = 1;
    // represents the hash of the pattern, stays constant after computing 
    int hash_p = 0;
    // the hash of the substring of s we are comparing against the pattern
    int hash_s = 0;

    // calculate hash_p and the initial hash_s
    for (int i = 0; i < len_p; i++) {
        hash_p = (d * hash_p + p[i]) % q;
        hash_s = (d * hash_s + s[i]) % q;
    }
    // calculating d^(len_p-1), we mod by q to prevent overflow. The math stays the same, as hash_s is also modded by q, 
    for (int i = 0; i < len_p - 1; i++) {
        d_p_1 = (d_p_1 * d) % q;
    }

    // loops through each possible substring
    for (int i = 0; i < len_s - len_p; i++) {
        // checks if hashes, and then strings are equal
        if (hash_p == hash_s && check_strings_equal(s,p,i)) {
            return i;
        }

        // first we remove the leftmost character from the string by substracting s[i] * d_p_1 from hash_s
        // then we add s[i + len_p] by performing our standard rolling hash equation h = (d * h + s[i]) % q;
        hash_s = (d * (hash_s - s[i] * d_p_1) + s[i + len_p] ) % q;
        // ensure the hash is positive
        if (hash_s < 0) {
            hash_s += q;
        }
    }

    // final check at the end of the string
    if (hash_p == hash_s && check_strings_equal(s,p,len_s - len_p)) {
        return len_s - len_p;
    }
    return -1;
}

int main()
{
    // s stores long string and p stores pattern
    string s;
    string p;

    // prompt the user for the long string
    getline(cin, s);

    // prompt the user for the search pattern
    getline(cin, p);

    // output the first index of 
    cout << rabin_karp(s, p);
    return 0;
}

File: projects/03-rabinKarp/implementations/rabinKarp.java
import java.util.Scanner;

public class rabinKarp {

    // The base we choose to use for our numbering system.
    // Since we are working with characters, we use 256, the number of distinct ASCII chars.
    public static final int d = 256;

    public static void main(String[] args){
        Scanner scanner = new Scanner(System.in);

        // Safely read the text from input
        if(!scanner.hasNextLine()) return;
        String text = scanner.nextLine();

        // Safely read the pattern from input
        if(!scanner.hasNextLine()) return;
        String pattern = scanner.nextLine();

        // Prime number used for modulo arithmetic.
        // This prevents integer overflows.
        int q = 101;

        // Execute the search, and output (print) the result.
        int result = search(pattern, text, q);
        System.out.println(result);

        scanner.close();
    }

    // This method takes in a string 'txt' and tries to find an inputted pattern 'ptrn' inside it.
    // It returns the first index of the first occurrence of the pattern in the text.
    // If no occurrence is found, it returns -1.
    static int search(String ptrn, String txt, int q){

        // Start by grabbing the lengths of the text and the pattern to find.
        // These will be used to construct our iterators.
        int m = ptrn.length();
        int n = txt.length();

        // If the pattern is longer than the text string, there is no way to find it inside.
        // Return false case (-1).
        if (m > n) return -1;

        // Initialize variables to hold the pattern and text window hashes, and a place value multiplier.
        int ptrnHash = 0;
        int txtHash = 0;
        int h = 1;

        // Calculate h = d^(m-1), which gives us the place value multiplier for the leftmost character in the window.
        for (int i = 0; i < m - 1; i++){
            h = (h * d) % q;
        }

        // Calculate the hash value of the pattern and the starting window of text.
        // hash = base * hash + character, for every character in the string.
        // Through this formula, we build the hash digits place by digits place.
        for (int i = 0; i < m; i++){
            ptrnHash = (d * ptrnHash + ptrn.charAt(i)) % q;
            txtHash = (d * txtHash + txt.charAt(i)) % q;
        }

        // Slide the window over the text, one character by one.
        // 'i' denotes the index of the start of the window.
        for (int i = 0; i <= n - m; i++){
            if (ptrnHash == txtHash){
                int j;
                // If hashes match, verify the literal string match character by character.
                // This rules out hash collisions.
                for (j = 0; j < m; j++){
                    if (txt.charAt(i + j) != ptrn.charAt(j)){
                        break;
                    }
                }

                // If j reached the length of the pattern, the match is perfect.
                // Return the first index of the match.
                // We return only the first index of the first match for consistency with the algorithm defined in the wiki.
                // Alternative implementations return the first index of all matches.
                if (j == m){
                    return i;
                }
            }

            // If the window has not yet reached the end of the text, we can slide it over one index and check again.
            // To do this, we first recalculate the hash for the next window.
            // This is done by removing the calculated addition to the hash for the first digit, then adding the value of the incoming one.
            // Mathematically, This removes the leading place value of the hash, slides the remaining digits left one, then adds the new least significant digit.
            if (i < n - m){
                txtHash = (d * (txtHash - txt.charAt(i) * h) + txt.charAt(i + m)) % q;

                // If the hash value is negative, make it positive.
                // This is equivalent to a single modulo operation.
                if (txtHash < 0){
                    txtHash = (txtHash + q);
                }
            }
        }
        // Pattern not found; return -1.
        return -1;
    }
}
